sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -56,36 +51,167 @@ public override void Process(HttpMessage message, ReadOnlyMemory<HttpPipelinePol
+                        }
+                        catch (AggregateException) { } // ignore exception here to rethrow it with EnsureCompleted
+                    }
+                }
 
-                _headerValue = "Bearer " + token.Token;
-                _refreshOn = token.ExpiresOn - TimeSpan.FromMinutes(2);
+                return headerValueTcs.Task.EnsureCompleted().HeaderValue;
             }
 
-            if (_headerValue != null)
+            private (TaskCompletionSource<HeaderValueInfo> tcs, TaskCompletionSource<HeaderValueInfo>? backgroundUpdateTcs, bool getTokenFromCredential) GetTaskCompletionSources()
             {
-                message.Request.SetHeader(HttpHeader.Names.Authorization, _headerValue);
+                lock (_syncObj)
+                {
+                    if (_infoTcs == null)
+                    {
+                        _infoTcs = new TaskCompletionSource<HeaderValueInfo>(TaskCreationOptions.RunContinuationsAsynchronously);
+                        return (_infoTcs, default, true);
+                    }
+
+                    if (!_infoTcs.Task.IsCompleted)
+                    {
+                        _backgroundUpdateTcs = default;
+                        return (_infoTcs, _backgroundUpdateTcs, false);
+                    }
+
+                    DateTimeOffset now = DateTimeOffset.UtcNow;
+                    if (_backgroundUpdateTcs != null && _backgroundUpdateTcs.Task.Status == TaskStatus.RanToCompletion && _backgroundUpdateTcs.Task.Result.ExpiresOn > now)
+                    {
+                        _infoTcs = _backgroundUpdateTcs;
+                        _backgroundUpdateTcs = default;
+                    }
+
+                    if (_infoTcs.Task.Status != TaskStatus.RanToCompletion || now >= _infoTcs.Task.Result.ExpiresOn)
+                    {
+                        _infoTcs = new TaskCompletionSource<HeaderValueInfo>(TaskCreationOptions.RunContinuationsAsynchronously);
+                        return (_infoTcs, default, true);
+                    }
+
+                    if (now >= _infoTcs.Task.Result.RefreshOn && _backgroundUpdateTcs == null)
+                    {
+                        _backgroundUpdateTcs = new TaskCompletionSource<HeaderValueInfo>(TaskCreationOptions.RunContinuationsAsynchronously);
+                        return (_infoTcs, _backgroundUpdateTcs, true);
+                    }
+
+                    return (_infoTcs, default, false);
+                }
             }
 
-            if (async)
+            private async ValueTask GetHeaderValueFromCredentialInBackgroundAsync(TaskCompletionSource<HeaderValueInfo> backgroundUpdateTcs, HeaderValueInfo info, HttpMessage httpMessage, bool async)
             {
-                await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
+                try
+                {
+                    HeaderValueInfo newInfo = await GetHeaderValueFromCredentialAsync(httpMessage, async);
```

Comment:
An interesting side effect of this implementation is that we would use current request's cancellation token for a background refresh.

Reply:
Yeah, but we don't have any better `CancellationToken`

Reply:
Is it better not to have any? Or have one with a baked in timeout (1 minute)?

Because current behavior is bad any way you look at it:
- If customer cancels the request they would cancel an unrelated token refresh too.
- If customer doesn't cancel background refresh would run unattended.

Reply:
Will be using `_tokenRefreshRetryDelay` for timeout, and run next retry without delay if this has failed on timeout